src/Microsoft.Azure.WebJobs.Host.Storage/StorageClientProvider.cs

```diff
@@ -0,0 +1,192 @@
+        private readonly AzureEventSourceLogForwarder _logForwarder;
+        private readonly ILogger _logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="StorageClientProvider{TClient, TClientOptions}"/> class that uses the registered Azure services.
+        /// </summary>
+        /// <param name="configuration">The configuration to use when creating Client-specific objects. <see cref="IConfiguration"/></param>
+        /// <param name="componentFactory">The Azure factory responsible for creating clients. <see cref="AzureComponentFactory"/></param>
+        /// <param name="logForwarder">Log forwarder that forwards events to ILogger. <see cref="AzureEventSourceLogForwarder"/></param>
+        public StorageClientProvider(AzureComponentFactory componentFactory, AzureEventSourceLogForwarder logForwarder, ILogger<StorageClientProvider<TClient, TClientOptions>> logger)
+        {
+            _componentFactory = componentFactory;
+            _logForwarder = logForwarder;
+            _logger = logger;
+
+            _logForwarder.Start();
+        }
+
+        /// <summary>
+        /// Gets the subdomain for the resource (i.e. blob, queue, file, table)
+        /// </summary>
+#pragma warning disable CA1056 // URI-like properties should not be strings
+        protected abstract string ServiceUriSubDomain { get; }
+#pragma warning restore CA1056 // URI-like properties should not be strings
+
+        /// <summary>
+        /// Gets the storage client specified by <paramref name="name"/>
+        /// </summary>
+        /// <param name="name">Name of the connection to use</param>
+        /// <param name="configuration">Configuration to use</param>
+        /// <param name="client">Client that was created if the connection was valid; otherwise, the default value for the type of <paramref name="client"/>.</param>
+        /// <returns>true if the client could be created; false otherwise</returns>
+        public virtual bool TryGet(string name, IConfiguration configuration, out TClient client)
+        {
+            try
+            {
+                client = Get(name, configuration);
+                return true;
+            }
+            catch (Exception ex)
+            {
+                client = default(TClient);
+                _logger.LogDebug("Could not create Storage Client. Exception: {0}", ex.ToString());
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Gets the storage client specified by <paramref name="name"/>
+        /// </summary>
+        /// <param name="name">Name of the connection to use</param>
+        /// <param name="resolver">A resolver to interpret the provided connection <paramref name="name"/>.</param>
+        /// <param name="configuration">Configuration to use</param>
+        /// <returns>Client that was created.</returns>
+        public virtual TClient Get(string name, INameResolver resolver, IConfiguration configuration)
```

Comment:
Was thinking that if INameResolver were a ctor param, you could get rid of this overload, and just run resolve over any names passed in.

Reply:
Would we need to define our own internal default INameResolver if we took a INameResolver in the ctor?